test/uk/gov/hmrc/merchandiseinbaggage/controllers/RetrieveDeclarationControllerSpec.scala

```diff
@@ -0,0 +1,94 @@
+
+  "onPageLoad" should {
+    s"return 200 with expected content" in {
+
+      val request = buildGet(routes.RetrieveDeclarationController.onPageLoad.url, aSessionId)
+      val eventualResult = controller(journey).onPageLoad(request)
+      val result = contentAsString(eventualResult)
+
+      status(eventualResult) mustBe 200
+      result must include(messageApi(s"retrieveDeclaration.title"))
+      result must include(messageApi(s"retrieveDeclaration.heading"))
+      result must include(messageApi(s"retrieveDeclaration.p"))
+
+      result must include(messageApi(s"retrieveDeclaration.mibReference.label"))
+      result must include(messageApi(s"retrieveDeclaration.mibReference.hint"))
+
+      result must include(messageApi(s"retrieveDeclaration.eori.label"))
+      result must include(messageApi(s"retrieveDeclaration.eori.hint"))
+    }
+  }
+
+  "onSubmit" should {
+    s"redirect to /previous-declaration-details after successful form submit" in {
+      val request = buildPost(routes.RetrieveDeclarationController.onSubmit().url, aSessionId)
+        .withFormUrlEncodedBody("mibReference" -> "XAMB0000010000", "eori" -> "GB123456780000")
+
+      val eventualResult = controller(journey).onSubmit(request)
+      status(eventualResult) mustBe 303
+      redirectLocation(eventualResult) mustBe Some(routes.RetrieveDeclarationController.onPageLoad().url)
+    }
+
+    "return 400 for invalid MibReference" in {
+      val request = buildPost(routes.RetrieveDeclarationController.onSubmit().url, aSessionId)
+        .withFormUrlEncodedBody("mibReference" -> "XAMB00000100", "eori" -> "GB123456780000")
+      val eventualResult = controller(journey).onSubmit(request)
+      val result = contentAsString(eventualResult)
+
+      status(eventualResult) mustBe 400
+      result must include(messageApi(s"retrieveDeclaration.mibReference.error.invalid"))
+    }
+
+    "return 400 for invalid Eori" in {
+      val request = buildPost(routes.RetrieveDeclarationController.onSubmit().url, aSessionId)
+        .withFormUrlEncodedBody("mibReference" -> "XAMB0000010000", "eori" -> "GB1234567")
+      val eventualResult = controller(journey).onSubmit(request)
+      val result = contentAsString(eventualResult)
+
+      status(eventualResult) mustBe 400
+      result must include(messageApi(s"retrieveDeclaration.eori.error.invalid"))
+    }
+
+    "return 400 for invalid form data" in {
+      val request = buildPost(routes.RetrieveDeclarationController.onSubmit().url, aSessionId)
+        .withFormUrlEncodedBody("mibReferzzz" -> "XAMB0000010000", "eorizszsz" -> "GB1234567")
+      val eventualResult = controller(journey).onSubmit(request)
+      val result = contentAsString(eventualResult)
+
+      status(eventualResult) mustBe 400
+      result must include(messageApi(s"retrieveDeclaration.mibReference.error.required"))
+      result must include(messageApi(s"retrieveDeclaration.eori.error.required"))
+    }
+  }
+}
```

Comment:
you have 3 tests here inside controller but zero Unit test for RetrieveDeclarationForm  - would be nice having 3 RetrieveDeclarationForm unit tests instead then only one controller asserting the wiring